### PR TITLE
[dashboard] Redirect to /blocked when trying to create or start a workspace while blocked

### PR DESCRIPTION
**Pull request by jankeromnes**
Fixes https://github.com/gitpod-io/gitpod/issues/3783

### CODE REVIEW

File: components/dashboard/src/start/CreateWorkspace.tsx [hunk 1]
```diff
@@ -75,16 +76,10 @@ export default class CreateWorkspace extends React.Component<CreateWorkspaceProp
...
-              search: `returnTo=${encodeURIComponent(window.location.toString())}&host=${error.data.host}&scopes=${error.data.scopes.join(',')}`
-          }).toString();
-          window.location.href = authorizeUrl;
-          statusMessage = <div className="mt-2 flex flex-col space-y-8">
-            <p className="text-base w-96">Redirecting to authorize with {error.data.host} …</p>
-            <a href={authorizeUrl}><button className="secondary">Authorize with {error.data.host}</button></a>
-          </div>;
-          break;
+          const authorizeURL = `/api/authorize?returnTo=${encodeURIComponent(window.location.toString())}&host=${error.data.host}&scopes=${error.data.scopes.join(',')}`;
+          return <Redirect push to={authorizeURL}/>;
```
Comment (jankeromnes) [inline components/dashboard/src/start/CreateWorkspace.tsx:80]: Interesting. This does point the address bar to the correct URL, but then the page is blank & doesn't do anything (I need to refresh for the authorization to work).

File: components/dashboard/src/start/CreateWorkspace.tsx [hunk 2]
```diff
@@ -75,16 +76,10 @@ export default class CreateWorkspace extends React.Component<CreateWorkspaceProp
...
+        case ErrorCodes.USER_BLOCKED:
+          return <Redirect to="/blocked"/>;
```
Comment (jankeromnes) [inline components/dashboard/src/start/CreateWorkspace.tsx:82]: Same here 😕

Screenshot 2021-04-07 at 09 21 00

Comment (jankeromnes) [reply components/dashboard/src/start/CreateWorkspace.tsx:82]: That's because `<Redirect/>` only works for `<Route/>`s in the same `<Switch/>`.

- `/api/authorize` is not a route
- `/blocked` is currently outside the `<Switch/>`

**Comment by jankeromnes** [comment]
How to test:

- Log in with **GitHub** OAuth at https://jx-redirect-blocked.staging.gitpod-dev.com/login
- Open a **GitLab** workspace, e.g. https://jx-redirect-blocked.staging.gitpod-dev.com/#https://gitlab.com/gitpod/spring-petclinic -- the GitLab OAuth redirect should be fast / barely noticeable (note: you may see a GitLab login/confirmation screen, or nothing GitLab-related if you had already authorized dev-staging before)
- Stop that workspace again from https://jx-redirect-blocked.staging.gitpod-dev.com/workspaces
- Next, open this PR in Gitpod: https://gitpod.io/#https://github.com/gitpod-io/gitpod/pull/3785
- Run `kubectl port-forward deployments/mysql 3306 &` and then `mysql -h 127.0.0.1 -p` (password is `test`)
- Run `use gitpod` and then `select id, name, blocked from d_b_user;` -- this will show your user ID
- Run `update d_b_user set blocked = 1 where id = '<ID>';` (where `<ID>` is your own user ID)
- Going back to https://jx-redirect-blocked.staging.gitpod-dev.com/workspaces, try to restart your old workspace, or to create a new workspace

In both cases, you should get quickly redirected to `/blocked`:

Screenshot 2021-04-09 at 14 00 14

**Comment by jankeromnes** [comment]
@gtsiolis or @AlexTugarev, please take a look if you have time 😊 (Priority: 🔅 Low)

File: components/dashboard/src/start/CreateWorkspace.tsx [hunk 3]
```diff
@@ -75,16 +75,8 @@ export default class CreateWorkspace extends React.Component<CreateWorkspaceProp
           </div>;
           break;
         case ErrorCodes.NOT_AUTHENTICATED:
-          const authorizeUrl = gitpodHostUrl.withApi({
-              pathname: '/authorize',
-              search: `returnTo=${encodeURIComponent(window.location.toString())}&host=${error.data.host}&scopes=${error.data.scopes.join(',')}`
-          }).toString();
-          window.location.href = authorizeUrl;
-          statusMessage = <div className="mt-2 flex flex-col space-y-8">
```
Comment (AlexTugarev) [inline components/dashboard/src/start/CreateWorkspace.tsx:83]: Please undo this change 🙏🏻 

It disagrees with https://github.com/gitpod-io/gitpod/issues/3754. I'm about to add handling for another error mode: user needs to select an account.

Comment (jankeromnes) [reply components/dashboard/src/start/CreateWorkspace.tsx:83]: I think we can probably have both changes (faster redirect, and /select-account logic).

But I'm happy to wait for your change and rebase/re-implement mine on top. 🙂

**Comment by AlexTugarev** [comment]
Screen Shot 2021-04-09 at 17 56 07

redirect on blocked state worked!

File: components/dashboard/src/start/CreateWorkspace.tsx (hunk 3 above)
Comment (AlexTugarev) [reply components/dashboard/src/start/CreateWorkspace.tsx:83]: @jankeromnes no we cannot, because we need to open the extra tab instead of the redirect to the auth flow. that extra tab is executing the auth flow and is going to propagate the "select account" error.

Comment (AlexTugarev) [reply components/dashboard/src/start/CreateWorkspace.tsx:83]: background: redirecting away from "this page" is always a risk of getting into a dead-end, so the user cannot find back.

Comment (jankeromnes) [reply components/dashboard/src/start/CreateWorkspace.tsx:83]: The `returnTo` URL seems to work generally well for that.

And, if there is indeed a rare/major complication in my auth flow, I think it's expected that I don't return to the exact same URL as before, right? (Because I need to deal with the bigger issue first, then later I can try again what I originally wanted to do.)

Comment (jankeromnes) [reply components/dashboard/src/start/CreateWorkspace.tsx:83]: But sure, I'll remove my change for now if it makes your more complicated.

Comment (jankeromnes) [reply components/dashboard/src/start/CreateWorkspace.tsx]: I've spliced my other commit into a separate branch: https://github.com/gitpod-io/gitpod/tree/jx/faster-start-auth-redirect

Next I'll wait for your `/select-account` change to re-evaluate.

Comment (AlexTugarev) [reply components/dashboard/src/start/CreateWorkspace.tsx]: It’s not about the details of the flow, but it is integrated into the SPA after all. We agreed to try the open in new window with good reasons and knowing that it would introduce this extra step for the benefit of more control of the overall experience. 

Such an extra step is rare, that’s true. So the idea is to support user’s instead of assuming that they know how oauth works.

Also this PR doesn’t sound like it would do this 😉

**Review by AlexTugarev — APPROVED** [review]
Works nicely! 🙏
